src/cephadm/cephadm

```diff
@@ -2047,24 +2070,70 @@ def default_image(func: FuncT) -> FuncT:
     return cast(FuncT, _default_image)
 
 
-def get_last_local_ceph_image(ctx: CephadmContext, container_path: str) -> Optional[str]:
+def get_container_info(ctx: CephadmContext, daemon_filter: str, by_name: bool) -> Optional[ContainerInfo]:
     """
+    :param ctx: Cephadm context
+    :param daemon_filter: daemon name or type
+    :param by_name: must be set to True if daemon name is provided
+    :return: Container information or None
+    """
+    def daemon_name_or_type(daemon: Dict[str, str]) -> str:
+        return daemon['name'] if by_name else daemon['name'].split('.', 1)[0]
+
+    if by_name and '.' not in daemon_filter:
+        logger.warning(f'Trying to get container info using invalid daemon name {daemon_filter}')
+        return None
+    daemons = list_daemons(ctx, detail=False)
+    matching_daemons = [d for d in daemons if daemon_name_or_type(d) == daemon_filter and d['fsid'] == ctx.fsid]
+    if matching_daemons:
+        d_type, d_id = matching_daemons[0]['name'].split('.', 1)
+        out, _, code = get_container_stats(ctx, ctx.container_engine.path, ctx.fsid, d_type, d_id)
+        if not code:
+            (container_id, image_name, image_id, start, version) = out.strip().split(',')
+            return ContainerInfo(container_id, image_name, image_id, start, version)
+    return None
+
+
+def infer_local_ceph_image(ctx: CephadmContext, container_path: str) -> Optional[str]:
+    """
+     Infer the local ceph image based on the following priority criteria:
+       1- the image specified by --image arg (if provided).
+       2- the same image as the daemon container specified by --name arg (if provided).
+       3- image used by any ceph container running on the host. In this case we use daemon types.
+       4- if no container is found then we use the most ceph recent image on the host.
+
+     Note: any selected container must have the same fsid inferred previously.
+
     :return: The most recent local ceph image (already pulled)
     """
+    # '|' special character is used to separate the output fields into:
+    #  - Repository@digest
+    #  - Image Id
+    #  - Image Tag
+    #  - Image creation date
     out, _, _ = call_throws(ctx,
                             [container_path, 'images',
                              '--filter', 'label=ceph=True',
                              '--filter', 'dangling=false',
-                             '--format', '{{.Repository}}@{{.Digest}}'])
-    return _filter_last_local_ceph_image(out)
-
+                             '--format', '{{.Repository}}@{{.Digest}}|{{.Id}}|{{.Tag}}|{{.CreatedAt}}'])
```

Comment:
This is causing a test failure. Possibly a docker vs. podman thing? The test used docker.
```
2022-04-22T01:52:38.648 INFO:teuthology.orchestra.run.smithi035.stderr:Inferring config /var/lib/ceph/8476b39c-c1dc-11ec-8c39-001a4aab830c/mon.smithi035/config
2022-04-22T01:52:38.701 INFO:teuthology.orchestra.run.smithi035.stderr:Non-zero exit code 1 from /usr/bin/docker images --filter label=ceph=True --filter dangling=false --format {{.Repository}}@{{.Digest}}|{{.Id}}|{{.Tag}}|{{.CreatedAt}}
2022-04-22T01:52:38.701 INFO:teuthology.orchestra.run.smithi035.stderr:/usr/bin/docker: stderr Template parsing error: template: :1:30: executing "" at <.Id>: can't evaluate field Id in type *formatter.imageContext
2022-04-22T01:52:38.702 INFO:teuthology.orchestra.run.smithi035.stderr:/usr/bin/docker: stderr
2022-04-22T01:52:38.708 INFO:teuthology.orchestra.run.smithi035.stderr:Traceback (most recent call last):
2022-04-22T01:52:38.708 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 9122, in <module>
2022-04-22T01:52:38.708 INFO:teuthology.orchestra.run.smithi035.stderr:    main()
2022-04-22T01:52:38.708 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 9110, in main
2022-04-22T01:52:38.708 INFO:teuthology.orchestra.run.smithi035.stderr:    r = ctx.func(ctx)
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 2035, in _infer_config
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:    return func(ctx)
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1980, in _infer_fsid
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:    return func(ctx)
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 2035, in _infer_config
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:    return func(ctx)
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 2060, in _infer_image
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:    ctx.image = infer_local_ceph_image(ctx, ctx.container_engine.path)
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 2133, in infer_local_ceph_image
2022-04-22T01:52:38.709 INFO:teuthology.orchestra.run.smithi035.stderr:    out, _, _ = call_throws(ctx,
2022-04-22T01:52:38.710 INFO:teuthology.orchestra.run.smithi035.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1739, in call_throws
2022-04-22T01:52:38.710 INFO:teuthology.orchestra.run.smithi035.stderr:    raise RuntimeError(f'Failed command: {" ".join(command)}: {s}')
2022-04-22T01:52:38.710 INFO:teuthology.orchestra.run.smithi035.stderr:RuntimeError: Failed command: /usr/bin/docker images --filter label=ceph=True --filter dangling=false --format {{.Repository}}@{{.Digest}}|{{.Id}}|{{.Tag}}|{{.CreatedAt}}: Template parsing error: template: :1:30: executing "" at <.Id>: can't evaluate field Id in type *formatter.imageContext
```
https://pulpito.ceph.com/adking-2022-04-22_01:18:37-orch:cephadm-wip-adk-testing-2022-04-21-1828-distro-basic-smithi/6799546/

Reply:
same error also hit in other test in that runhttps://pulpito.ceph.com/adking-2022-04-22_01:18:37-orch:cephadm-wip-adk-testing-2022-04-21-1828-distro-basic-smithi/6799587/